### DISCUSSION TITLE
ESP32C3 + ST7735S: The black fill of the display doesn't work correctly

### DISCUSSION THREAD

**Discussion by PospelovDaniil** [Bodmer/TFT_eSPI#2768] [New processor/display support or capability request]
Hi Bodmer.
I have a problem with the ST7735S display.

1. The black fill of the display doesn't work correctly - there are different colored stripes on the edges, 1-2 pixels wide 
2. Arduino IDE
3. TFT_eSPI library version 2.5.0
4. esp32 2.0.9
5. ESP32C3 [Luat OS](https://aliexpress.ru/item/1005004428915044.html?spm=a2g2w.orderdetail.0.0.78194aa6retGza&sku_id=12000029153808053)
6. TFT driver ST7735S
7. Interface type SPI 

My user setup:
```
// Setup for ESP32 and ST7735 80 x 160 TFT
#define USER_SETUP_ID 43
// See SetupX_Template.h for all options available

#define ST7735_DRIVER


#define TFT_WIDTH  80
#define TFT_HEIGHT 160


#define ST7735_GREENTAB160x80

// For ST7735, ST7789 and ILI9341 ONLY, define the colour order IF the blue and red are swapped on your display
// Try ONE option at a time to find the correct colour order for your display

//  #define TFT_RGB_ORDER TFT_RGB  // Colour order Red-Green-Blue
//  #define TFT_RGB_ORDER TFT_BGR  // Colour order Blue-Green-Red

#ifdef ESP32
#define TFT_MISO 3//19
#define TFT_MOSI 3//23
#define TFT_SCLK 2//18
#define TFT_CS    7//15  // Chip select control pin
#define TFT_DC    6//2  // Data Command control pin
#define TFT_RST   10//4  // Reset pin (could connect to RST pin)
//#define TFT_RST  -1  // Set TFT_RST to -1 if display RESET is connected to ESP32 board RST
#else
// Display GND       to NodeMCU pin GND (0V)
// Display VCC       to NodeMCU 5V or 3.3V
// Display SCK       to NodeMCU pin D5
// Display SDI/MOSI  to NodeMCU pin D7
// Display BLK       to NodeMCU pin VIN
#define TFT_CS   PIN_D8  // Chip select control pin D8
#define TFT_DC   PIN_D3  // Data Command control pin
#define TFT_RST  PIN_D4  // Reset pin (could connect to NodeMCU RST, see next line)
#endif

#define LOAD_GLCD   // Font 1. Original Adafruit 8 pixel font needs ~1820 bytes in FLASH
#define LOAD_FONT2  // Font 2. Small 16 pixel high font, needs ~3534 bytes in FLASH, 96 characters
#define LOAD_FONT4  // Font 4. Medium 26 pixel high font, needs ~5848 bytes in FLASH, 96 characters
#define LOAD_FONT6  // Font 6. Large 48 pixel font, needs ~2666 bytes in FLASH, only characters 1234567890:-.apm
#define LOAD_FONT7  // Font 7. 7 segment 48 pixel font, needs ~2438 bytes in FLASH, only characters 1234567890:.
#define LOAD_FONT8  // Font 8. Large 75 pixel font needs ~3256 bytes in FLASH, only characters 1234567890:-.
//#define LOAD_FONT8N // Font 8. Alternative to Font 8 above, slightly narrower, so 3 digits fit a 160 pixel TFT
#define LOAD_GFXFF  // FreeFonts. Include access to the 48 Adafruit_GFX free fonts FF1 to FF48 and custom fonts

#define SMOOTH_FONT


//#define SPI_FREQUENCY  20000000
  #define SPI_FREQUENCY  27000000 // Actually sets it to 26.67MHz = 80/3

```
Example code:
```
#include <TFT_eSPI.h>
static TFT_eSPI tft;
void setup()
{
  tft.init();
  tft.invertDisplay(false);
  tft.setRotation(3);
  tft.fillScreen(TFT_BLACK);
  tft.setTextSize(3);
  tft.setTextColor(TFT_RED);
  tft.printf("my text");
}

void loop()
{
}

```

![L0AP07X6TUE](https://github.com/Bodmer/TFT_eSPI/assets/36377393/8c6ddf48-8e41-4364-8fee-7525acacdb42)

**Comment by PospelovDaniil** [comment]
Method of fixing: enable erasing the entire flash before loading the script.
Can you explain why this helped and why without it this effect?
You can close the issue

**Comment by PospelovDaniil** [comment]
The fix has stopped working

**Comment by Bodmer** [comment]
This problem occurs because there is an offset of the 160x80 display zone within the maximum resolution area of the driver chip (132x162).

You could try using:
```
#define INITR_REDTAB160x80
```
Instead of:
```
#define ST7735_GREENTAB160x80
```

If that causes problems with colours, then it is necessary to edit the offsets here:
https://github.com/Bodmer/TFT_eSPI/blob/master/TFT_Drivers/ST7735_Rotation.h#L25

f you do not want to modify the library, then the other option is to define the display as full resolution and use the setViewport function to compensate for the offset.

Eventually I will add a new parameter to allow users to change the offsets.

**Comment by PospelovDaniil** [comment]
1) I tried replacing ST7735_GREENTAB160x80 with INITR_REDTAB160x80 - the result was no text
![image](https://github.com/Bodmer/TFT_eSPI/assets/36377393/5761738e-4cf7-4d00-9f6c-ed8e732a5fd0)

2) I returned ST7735_GREENTAB160x80 and replaced the width and height:
#define TFT_WIDTH  132
#define TFT_HEIGHT 162
![image](https://github.com/Bodmer/TFT_eSPI/assets/36377393/cd2593f6-f550-4ab6-be73-f97f65fa7ab7)

**Comment by PospelovDaniil** [comment]
replacing with
#define TFT_WIDTH 162
#define TFT_HEIGHT 132
gives the same result as on the 2nd picture

**Comment by Bodmer** [comment]
Use the setViewport function to compensate for the offset. You can determine the offsets by experimentation.
